### PR TITLE
sbcl-bootstrap: update powerpc boot binary

### PR DESCRIPTION
**Pull request by barracuda156**
#### Description

Update later, since 2.4.6 fails on macOS 13. Here update powerpc boot binary, which is required to support powerpc now (2.4.4 has been broken, since the old boot binary cannot build it).

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.6
Xcode 3.2

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by barracuda156** [comment]
@catap Any idea why macOS-13 build fails here?
```
  Argh! corrupted error depth, halting
  fatal error encountered in SBCL pid 4871 pthread 0x7ff8579fd7c0:
  %PRIMITIVE HALT called; the party is over.
  
  Error opening /dev/tty: Device not configured
  Welcome to LDB, a low-level debugger for the Lisp runtime environment.
  ldb> 
  Command failed:  cd "/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_lang_sbcl/sbcl/work/sbcl-2.4.6" && sh make.sh --prefix=/opt/local --xc-host=/opt/local/libexec/sbcl-bootstrap/bin/sbcl --arch=x86-64 --without-sb-simd --fancy 
  Exit code: 1
  Error: Failed to build sbcl: command execution failed
  DEBUG: Error code: CHILDSTATUS 4173 1
```

**Comment by catap** [comment]
@barracuda156 everything is crashed seems quite deep inside:
```
Argh! corrupted error depth, halting
  fatal error encountered in SBCL pid 4871 pthread 0x7ff8579fd7c0:
  %PRIMITIVE HALT called; the party is over.
```
and after that it simple can't run debugger because no tty.

**Comment by barracuda156** [comment]
@catap What should we do about it? Ask in the mailing list?

**Comment by catap** [comment]
On Fri, 05 Jul 2024 09:33:01 +0100,
Sergey Fedorov ***@***.***> wrote:
> 
> @catap What should we do about it? Ask in the mailing list?
>

Indeed

-- 
wbr, Kirill

**Comment by barracuda156** [comment]
I will drop sbcl update first. Let us deal with powerpc boot binary first.

**Comment by barracuda156** [comment]
@herbygillot Could you help with merging this? I have dropped an update to SBCL as such, what is left is a needed update of PowerPC boot binary. It only affects ppc systems, where at the moment SBCL does not build due to outdated boot binary.

File: lang/sbcl-bootstrap/Portfile
Comment (pmetzger) [inline lang/sbcl-bootstrap/Portfile:79]: Whose site is this? It's not the SBCL project.

Comment (catap) [reply lang/sbcl-bootstrap/Portfile:79]: I control this domain and server for it, and Sergey fills it with data.

Comment (pmetzger) [reply lang/sbcl-bootstrap/Portfile:79]: That makes it difficult to trace the provenance of such bootstrap binaries etc.

Comment (catap) [reply lang/sbcl-bootstrap/Portfile:79]: Do you have better idea? SBCL can't be bootstrap by so old SBCL anymore.

Comment (pmetzger) [reply lang/sbcl-bootstrap/Portfile:79]: Can it be bootstrapped from a Common Lisp interpreter in C? CLISP say?

Comment (catap) [reply lang/sbcl-bootstrap/Portfile:79]: The last time I tried it can't. Thus, I not sure that modern enough Clisp or CCL is supported by PowerPC as well.

Comment (pmetzger) [reply lang/sbcl-bootstrap/Portfile:79]: Is the problem that you can't compile clisp for powerpc? It's pure C.

Comment (catap) [reply lang/sbcl-bootstrap/Portfile:79]: The problem that SBCL can't be bootstraped on macOS by CLisp. I've tried it a year or two ago last time and it had filed. I had opened a few bugs in upstream but nobody reacted.

Comment (pmetzger) [reply lang/sbcl-bootstrap/Portfile:79]: I think I could light a fire under upstream if necessary.

Comment (pmetzger) [reply lang/sbcl-bootstrap/Portfile:79]: (BTW, what specifically fails?)

Comment (barracuda156) [reply lang/sbcl-bootstrap/Portfile:79]: @pmetzger Perry, this is a temporary measure, which was suggested by SBCL upstream. They do not have powerpc hardware, and have to rely on us.

There is no requirement for this to be a binary provided by myself and Kirill though. If @kencu can help to compile it on his powerpc machine and upload it somewhere to MacPorts hosting, that would be even better. We need a standard build here, which upstream can reasonably trust.

Comment (pmetzger) [reply lang/sbcl-bootstrap/Portfile:79]: I think it's better to bootstrap it off of CLISP. In theory they're both ANSI Common Lisp and it should be possible. What's needed is detail on how and why this is failing.

Comment (barracuda156) [reply lang/sbcl-bootstrap/Portfile:79]: @ryandesign Ryan, any chance of having a limited (controlled by you) buildbot for powerpc set up? That would have solved the problem, though we would need a two-stage bootstrap for powerpc for a while.

Comment (pmetzger) [reply lang/sbcl-bootstrap/Portfile:79]: I really still think it's better to boostrap off CLISP. There are probably bugs right now and upstream is almost certainly interested in fixing them. I know enough lispheads that I could make the right contacts.

Comment (pmetzger) [reply lang/sbcl-bootstrap/Portfile:79]: CLISP is portable and written in C. It does not include a compiler, it's a pure interpreter, and upstream is responsive.

Comment (barracuda156) [reply lang/sbcl-bootstrap/Portfile:79]: If I understood Kirill correctly, the problem is not to build Clisp, but to use it to build SBCL.

@catap Could you refer us to the issues with Clisp on macOS? Even regardless of this ticket, it will be nice to have those fixed.

**Comment by barracuda156** [comment]
@catap Kirill, could you make a variant of the portfile which uses `clisp` for the bootstrap? I am not asking you to test it, just to share what you have used yourself. `clisp` builds on `ppc`, FWIW, so trying does not hurt.

**Comment by catap** [comment]
> @catap Kirill, could you make a variant of the portfile which uses `clisp` for the bootstrap? I am not asking you to test it, just to share what you have used yourself. `clisp` builds on `ppc`, FWIW, so trying does not hurt.

Such variant can be implemented quite strigth forward and similar with `ecl`. You need just use clisp as dependency and bootstrap lisp.

I haven't recall how it fails, but it was something inside clisp but I only tried it on i386 or x86_64.

File: lang/sbcl-bootstrap/Portfile
Comment (pmetzger) [reply lang/sbcl-bootstrap/Portfile:79]: clisp should be able to build SBCL, but the maintainer doesn't really have a modern CI for all this stuff and is probably not regularly testing it. But he'll probably take patches.

**Comment by pmetzger** [comment]
@catap We should try to fix it for all platforms. I'm pretty sure upstream will not mind taking patches etc.

**Comment by barracuda156** [comment]
> @catap We should try to fix it for all platforms. I'm pretty sure upstream will not mind taking patches etc.

Well, provided someone fixes it for `i386`, I can sort out `ppc`.

@catap By the way: https://gitlab.com/gnu-clisp/clisp/-/issues/52

**Comment by pmetzger** [comment]
If someone gives me a portfile for x86_64 that uses clisp for bootstrap I can run it and have a look at the errors.

**Comment by reneeotten** [comment]
so this builds on modern CI systems. Can we just merge this and then interested parties can deal with whatever issues remain when they have time?

**Comment by reneeotten** [comment]
oh, actually this only touches PPC. So does that work now or not? If so, and it doesn't break anything else, then let's just merge this - or?

**Comment by barracuda156** [comment]
> oh, actually this only touches PPC. So does that work now or not? If so, and it doesn't break anything else, then let's just merge this - or?

@reneeotten It is ppc-only, affects nothing else and would allow us to have ppc fixed. Upstream recommended doing it this way, and then they can pick our binary to host on their side. Once that is done, we change to using that.

**Comment by pmetzger** [comment]
Being able to bootstrap without a binary common lisp is better. As I said, I'll find the problem in sbcl + clisp for bootstrap if someone will get me a reproducer.

**Comment by barracuda156** [comment]
> Being able to bootstrap without a binary common lisp is better. As I said, I'll find the problem in sbcl + clisp for bootstrap if someone will get me a reproducer.

Well, since the build of `clisp` itself is broken and upstream ignores the issue, I can’t do much.

`ecl` is broken for all 32-bit apps platforms: https://gitlab.com/embeddable-common-lisp/ecl/-/issues/705#note_1865559254

**Comment by pmetzger** [comment]
What is wrong with clisp?

**Comment by barracuda156** [comment]
> What is wrong with clisp?

https://gitlab.com/gnu-clisp/clisp/-/issues/52

**Comment by pmetzger** [comment]
That problem with clisp looks pretty easy to fix.

**Comment by reneeotten** [comment]
okay, so @barracuda156 @pmetzger what do you want to do with this PR. There are too many tangential discussion going on here that appear to be possible to address later (or never) but do not directly affect this PR - correct? If so, let's merge this otherwise I am going to close this as there has been no progress in months.

**Comment by pmetzger** [comment]
I said: 

> As I said, I'll find the problem in sbcl + clisp for bootstrap if someone will get me a reproducer.

I was then told the real problem is clisp doesn't build on old systems, which seems strange to me and I asked what was wrong with it, and I got pointed to something that seems to show a trivial redefinition of alignof is happening that can probably be patched pretty easily. @barracuda156 should probably just patch that problem and then bootstrap sbcl off clisp.

**Comment by barracuda156** [comment]
@pmetzger First of all, building with clisp is a completely unrelated issue. Yes, it is desirable to be able to bootstrap from C. I support that. It is not desirable to force everyone to bootstrap from C though. It is also not a recommended method by SBCL upstream.

Building with clisp should be implemented for all platforms, not just for powerpc, and in the separate PR.

Here I would like to update bootstrap binary, because existing one does not work for new versions, and upstream advised me and Kirill to do this.

I do not get why this should be distracted into clisp issue, however desirable it is on its own.

@reneeotten It should be merged, because it fixes powerpc without touching anything else. I do not get why all of a sudden so much interest to ppc specifically here, but once someone has a better solution, it is always possible to change portfile accordingly.

**Comment by easye** [comment]
I thought that the problem with merging the PR was that the PPC binary bootstrap was not to be hosted by the SBCL developers, or does no one have an issue with this?  I believe that the SBCL developers consider PPC to be no longer actively supported, so they don't want to host a binary release that they won't answer questions/takes bugs against.  

If the MacPorts org doesn't have a problem with the PPC bootstrap image originating from a non-standard download location, I would then agree we should just merge this patch.  

In practice, I don't think that such usage of a non-standard download location is any less secure than the current download from SourceForge (where sbcl.org hosts their downloads).  Arguably it might take less effort to put a malicious payload in the binary on SourceForge than in the "non-standard" download location.  I don't have a great feel for what "MacPorts" wants to do here…

**Comment by catap** [comment]
@easye `macos-powerpc.org` is mine domain which I control and which I prepared with a server to host artifacts for @barracuda156.

The scope of persons who has access to content on this machine is limited for 2 persons: me and @barracuda156.

Future of that, provided binary has it's signature inside this PR and if it was changed, the signature will be missmatched.

Base on amount of effort from @barracuda156 to support powerpc I really doubt that his position can be questionable. If it is quite complicated and very long an attack for supply chain... well... it has quite limited expose: a handful of guys who uses macports on powerpc. I really doubt that it is hunder of pepole, quite possible that less than a few dozens.

**Comment by easye** [comment]
@catap I am in favor of your hosting the binaries, mainly because sbcl.org never will (due to lack of resources to support PPC from what I understand).

I would just like a statement from someone "higher up" in the MacPorts organization (or something pointing to written down policy), that use of such a repository in this case is acceptable.

**Comment by barracuda156** [comment]
@easye That would be really weird clause explicitly stating that a repo owned by a MacPorts contributor is allowed. How does it differ in any meaningful way from a repo owned by someone else?
FWIW, I think some Rust ports use/used a personal GitHub to host binaries. There is really no other way as long as a) binaries are required and b) upstream does not happen to host them for w/e reason.

**Comment by easye** [comment]
@barracuda156 I am just looking for some (any?) guidance on the incorporation of repos in MacPorts.  I agree with you that there is little meaningful difference on who "owns" a repo other than some intangible qualia on whether one "trusts" this or that.  Given that we are in the era of long running state actor-resourced attacks on supply chains, such concerns are less "academic" than before, but probably not any better reasoned.

File: lang/sbcl-bootstrap/Portfile
Comment (easye) [reply lang/sbcl-bootstrap/Portfile:79]: After much time having passed, and having noted that SBCL maintainers have made a conditional offer to host the PPC binaries in the futures once MacPorts takes a step towards earning trust <https://sourceforge.net/p/sbcl/mailman/message/58774549/>, I am for merging this PR as-is, over @pmetzger's preference for a clisp bootstrap mechanism.  Were someone to produce patches to use CLISP (or ABCL) to bootstrap, I would sponsor those patches.

**Review by easye — APPROVED** [review]
Some PPC support is better than none.

**Comment by easye** [comment]
Accepting this pull request:  apologies about the elapsed time.

**Comment by barracuda156** [comment]
@easye Thank you very much!

**Comment by pmetzger** [comment]
I was away for a couple of days. I'm pretty irritable about this. This should be bootstrapped from clisp, not built from an opaque binary.

**Comment by easye** [comment]
Sorry to have missed the opportunity for you to reaffirm your opposition here, and consider it more strongly before merging.  

A quick summary of why I went ahead and merged:

1.  {{{sbcl}}} is currently bootstrapped from opaque binaries provided sbcl.org.  @pmetzger Do you desire to rework the {{{Portfile}}} not to use those binaries at all, i.e. is your opposition more about using binaries to build a port than about the provenance of said binaries?  In that case, there should be no {{{sbcl-bootstrap}}} port at all.

2.  I see @catap as at least as trustworthy a source as {{{sbcl.org}}} of opaque binaries.   The Portfile still requires a valid checksum from whatever source it uses.   

3.  {{{sbcl.org}}} has no resources to support PPC, but has indicated that they would consider hosted PPC binaries if we were to get the port working on PPC, thus this is hopefully the first step in single-sourcing the build binaries.

4.  This only effects the PPC builds of SBCL, for which I think @barracuda156 is really the only user, so the impact on almost all other users of {{sbcl-bootstrap}}} is quite small.

**Comment by pmetzger** [comment]
My point is we do not need to be bootstrapping from binaries. We have perfectly good common lisp ports that can be built from sources, and those were used in the past to bootstrap SBCL.

Also, I do not like "trusting" anyone; see the recent "xz" incidents. It should be possible for people to look at code and tell if it's okay, we should not need to "trust".

**Comment by barracuda156** [comment]
@pmetzger This is a great suggestion, and I hope someone starts here with something more significant and impactful than a version of Lisp on an archaic platform with a handful of users. For example, bootstrapping Rust from C would contribute far more to security, and then perhaps bootstrapping Haskell from C. I am not referring to powerpc here, of course: implement bootstrapping these two on x86_64 and arm64. Also stop using opaque prebuilt binaries with Go and bootstrap it always from C via Go-1.4. Many more things to improve here, which will benefit users of modern systems on supported archs.

**Pull request merged**
